src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+    index_id: String,
+
+    pub(crate) findex_master_key: KeyingMaterial<MASTER_KEY_LENGTH>,
+
+    fetch_entries_key: Option<[u8; 16]>,
+    fetch_chains_key: Option<[u8; 16]>,
+    upsert_entries_key: Option<[u8; 16]>,
+    insert_chains_key: Option<[u8; 16]>,
+}
+
+impl Display for Token {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut keys = self.findex_master_key.to_vec();
+
+        if let Some(fetch_entries_key) = self.fetch_entries_key {
+            keys.push(0);
+            keys.extend(fetch_entries_key);
+        }
+
+        if let Some(fetch_chains_key) = self.fetch_chains_key {
+            keys.push(1);
+            keys.extend(fetch_chains_key);
+        }
+
+        if let Some(upsert_entries_key) = self.upsert_entries_key {
+            keys.push(2);
+            keys.extend(upsert_entries_key);
+        }
+
+        if let Some(insert_chains_key) = self.insert_chains_key {
+            keys.push(3);
+            keys.extend(insert_chains_key);
+        }
+
+        write!(f, "{}{}", self.index_id, base64::encode(keys))
+    }
+}
+
+impl FromStr for Token {
+    type Err = FindexErr;
+
+    fn from_str(token: &str) -> Result<Self, Self::Err> {
+        let (index_id, tail) = token.split_at(5);
+        let mut bytes = base64::decode(tail)
+            .map_err(|_| {
+                FindexErr::Other(format!(
+                    "token {token} is not a valid base64 encoded string"
+                ))
+            })?
+            .into_iter();
+
+        let findex_master_key =
+            KeyingMaterial::try_from_bytes(&bytes.next_chunk::<16>().map_err(|_| {
+                FindexErr::Other(
+                    "the token is too short, cannot read the Findex master key".to_owned(),
+                )
+            })?)?;
+
+        let mut token = Token {
+            index_id: index_id.to_owned(),
+            findex_master_key,
+
+            fetch_entries_key: None,
+            fetch_chains_key: None,
+            upsert_entries_key: None,
+            insert_chains_key: None,
+        };
+
+        while let Some(prefix) = bytes.next() {
+            let key = Some(bytes.next_chunk::<16>().map_err(|_| {
+                FindexErr::Other(format!(
+                    "the token is too short, expecting 16 bytes after the prefix {prefix}"
+                ))
+            })?);
+
+            if prefix == 0 {
+                token.fetch_entries_key = key;
+            } else if prefix == 1 {
+                token.fetch_chains_key = key;
+            } else if prefix == 2 {
+                token.upsert_entries_key = key;
+            } else if prefix == 3 {
+                token.insert_chains_key = key;
+            } else {
+                return Err(FindexErr::Other(format!(
+                    "the token contains a unknown prefix {prefix}"
+                )));
+            }
+        }
+
+        Ok(token)
+    }
+}
+
+impl Token {
+    pub fn random_findex_master_key(
```

Comment:
Yes, we should leave dead code.

Moreover I don't see why you would need to use this function. To my understanding, the master key should be generated by the master authority when a reindexing is done while the token is generated when a user requests it. It seems these are two distinct operations. The master key being generated before a user asks for it.

I don't know when the callback keys should be generated, but I guess it should be something similar.

Plus, you should take `Option`s of your callback keys as input.

Reply:
`random_findex_master_key` is used in wasm and will be used inside FFI next.

We need this function to create the initial token. When creating an index in the console the backend will return all the signing callbacks keys, but the user need to create the token from these key (adding the random findex master key). We could have two operations (one to generate the random findex master key and one to generate the token from all the keys) but it's simpler this way since both operations are done together.

Reply:
Why does the user create a random key ? I think we need to keep these two operation separated as I don't understand why we couldn't want to create a token without generating a new random key.

Reply:
Because the user need the Findex Master Key for the new index? The backend cannot generate this key since it's a private key.

Right now, we don't want to create a token without generating a new random key, let's keep it simple for clients (the console for exemple) with only one call to wasm, if we need to separate in the future we could create the two distinct operations and keep this one as an alias.

Reply:
ok

Reply:
Actually, I think it should be cleaner to generate the master key in the interface code. The token would still be initialized with a new random key upon a unique wasm call but the Rust code shouldn't mix these concepts imho.

Reply:
It's hard to have a correct random generator in all languages (for example in JS `crypto` is different in NodeJS and browser so it require some hard packaging to allow both target, having it in Rust is simpler)

Reply:
I mean you could still generate the randomness in Rust, but inside the WASM interface (for example). This would give something like:
wasm_call -------> rust_wasm_interface --------------> generate random key
                                                                     |
                                                                      --------------> generate token

Reply:
I don't want to have copy/paste inside FFI/Wasm if it's the same in both, no?

Reply:
This is just two lines. I would do it, but I understand your point. Maybe we can ask @Manuthor to give us his opinion?

Reply:
I would keep the random generation in Rust (can be tricky to have good implementations in every language).

I agree that the 2 operations should be done separately (master key generation and authentication keys generation). But, for convenience, I agree that this token should be also self-sufficient (and then generated in one pass).

But one question: since master key is renewed at each call, how do you deliver multiple tokens with the same Findex master key?